README
NAME
    Math::Factor::XS - Factorize numbers and calculate matching
    multiplications

SYNOPSIS
     use Math::Factor::XS ':all';
     # or
     use Math::Factor::XS qw(factors prime_factors matches);

     $number = 30107;

     @factors = factors($number);
     @primes = prime_factors($number);
     @matches = matches($number, \@factors);

     print "$factors[1]\n";
     print "$number == $matches[0][0] * $matches[0][1]\n";

DESCRIPTION
    "Math::Factor::XS" factorizes numbers by applying trial divisions.

FUNCTIONS
  factors
    Find all factors (ie. divisors) of a number.

     @factors = factors($number);

    The number is factorized and its factors are returned as a list. For
    example,

     @factors = factors(30);
     #  @factors = (2, 3, 5, 6, 10, 15);

  prime_factors
    Find prime factors of a number.

     @factors = prime_factors($number);

    The number is factorized and its prime factors are returned as a list.
    Multiplying the list together gives $number. For example,

     @primes = prime_factors(90);
     #  @primes = (2, 3, 3, 5);

  matches
    Calculates matching multiplications.

     @matches = matches($number, \@factors, { skip_multiples => [0|1] });

    The factors will be multiplied against each other and all combinations
    that equal the number itself will be returned as a two-dimensional list.
    The matches are accessible through the indexes; for example, the first
    two numbers that matched the number may be accessed by $matches[0][0]
    and $matches[0][1], the second pair by $matches[1][0] and
    $matches[1][1], and so on.

    The hashref provided at the end is optional. If "skip_multiples" is set
    to a true value, then matching multiplications that contain
    multiplicated small factors will be discarded. Example:

     11 * 2737 == 30107 # accepted
     77 * 391  == 30107 # discarded

    Direct use of $Math::Factor::XS::Skip_multiple does no longer have an
    effect as it has been superseded by "skip_multiples".

EXPORT
  Functions
    "factors()", "matches()" and "prime_factors()" are exportable.

  Tags
    ":all - *()"

AUTHOR
    Steven Schubiger <schubiger@cpan.org>

LICENSE
    This program is free software; you may redistribute it and/or modify it
    under the same terms as Perl itself.

    See <http://dev.perl.org/licenses/>



Changes
Revision history for Perl extension Math::Factor::XS.

0.39_01  2012-08-01  <kryde@cpan.org>

 - factors() range check just with SvNV, so as to accept overloaded
   types such as Math::BigInt (if fit ulong) and for big speedup.
 - New count_prime_factors().

0.39  2012-01-19  <kryde@cpan.org>

 - Stable release.

0.38_03  2012-01-08  <kryde@cpan.org>

 - prime_factors() step by 2,4 to skip multiples of 3

0.38_02  2011-12-30  <kryde@cpan.org>

 - Use Perl's TRUE/FALSE to avoid trouble with enum{false,true} on
   systems such as BSD which drag in stdbool.h

0.38_01  2011-12-28  <kryde@cpan.org>

 - New prime_factors().
 - Clarify docs of factors() that it's all divisors.

0.38  2010-03-26  <schubiger@cpan.org>

 - Merged development version to stable.

0.37_02  2010-03-24  <schubiger@cpan.org>

 - Allow validation of factors to succeed when none are supplied.
   [rt #53739 - Tom Scheper]

 - Return an empty list from xs_matches() if there are no factors.

 - Add a test for no factors provided to matches().

 - Don't force explicit scalar context where an implicit one exists.

0.37_01  2010-03-01  <schubiger@cpan.org>

 - Fix memory leak in xs_factors(). [reported by William P.N. Smith]

0.37 Tue Nov 10 22:37:24 CET 2009

 - Merged development version to stable.

0.36_01 Fri Nov  6 10:17:32 CET 2009

 - Save the quotient as factor only when it's greater than the iterator.
   [rt #50999 - Thor Andreassen]

 - Add a test for even numbers to be factorized.

 - Check the numbers returned deeply against the expected ones.

 - Update broken license links.

0.36 Tue Sep 29 11:36:53 CEST 2009

 - Merged development version to stable.

0.35_02 Tue Sep 15 20:05:35 CEST 2009

 - Check up to square root of a number to determine its factors.
   [rt #46870 - Jonathan Leto]

 - Declare the reference of the factors array which is passed to
   xs_matches() as SV * (instead of AV *). This fixes an error
   when running under perl 5.6.2.

0.35_01 Fri Jun 12 16:35:25 CEST 2009

 - Fix Build.PL related test failures. [rt #39211 - Slaven Rezic]

 - Using the global 'Skip_multiple' does no longer have an effect;
   passing the factors to matches() requires now a list reference.

 - Validation of arguments is handled by Params::Validate.

 - Use printf() instead of formats within the example script.
   This circumvents a segmentation fault reported in [perl #64562].

 - The options hash for the matches() function is being passed to
   the according XSUB as reference.

0.35 Sun Feb  3 14:40:33 CET 2008

 - Reworded documentation.

 - Removed unneeded code within scripts/factors_matches.pl.

 - Renamed the calc_factors.t test-file to the more generic functions.t.

 - Updated ppport.h.

0.34 Tue Aug 14 18:40:38 CEST 2007

 - Beautified documentation (added code tags where suitable).

0.33 Sat Apr  1 16:47:21 CEST 2006

 - Changed Build.PL argument create_makefile_pl from 'passtrough'
   to 'traditional' for the sake of compatibility.

0.30 Fri Nov  4 20:52:13 CET 2005

 - Converted Math::Factor over to XS.

0.29 2005/09/21

 - factors() takes a single number and returns now a flat array
   of factors. matches() likewise. [Ricardo Signes, <rjbs@cpan.org>]

 - Removed each_factor() & each_match() due to data that
   is no longer preserved in multidimensional hash structures
   and therefore doesn't require these accessors.

0.19 2004/01/19

 - Formats are not evaled in loops.

0.17 2004/01/18

 - Matching multiplications that contain smaller factors may
   be skipped.

0.16 2004/01/17

 - factor.pl revisioned.

0.12 2004/01/13

 - Hashes are returned as refs.

0.1 2004/01/12

 - Minor additions.

0.09 2004/01/12

 - Evaluation of ranges implemented.

0.07 2004/01/10

 - show_factor() & show_match() have been moved out of the
   package to bin/factor.pl.

0.05 2004/01/07

 - Iteration over factors and multiplication matches by
   each_factor() & each_match().

0.03 2004/01/06

 - Heredocs mostly replaced with format in show_factor()
   & show_match().

0.02 2004/01/06

 - Documentation covers functionality.

0.01 2004/01/06

 - Initial version.


scripts/usage-example.pl
#!/usr/bin/perl

use strict;
use warnings;

use Math::Factor::XS qw(factors matches);

our (%form, $i, $matches, $ul);

#$Math::Factor::XS::Skip_multiple = 1;
    
my $number = 30107;

my @factors = factors($number);
my @matches = matches($number, @factors);

show_factors($number, \@factors);
show_matches($number, \@matches);

sub show_factors {
    my ($number, $factors) = @_;

    print <<'HEADER';
-------
factors
-------

HEADER
     
    local $ul = '-' x length($number);
	
    formeval('factors');
    write; 
    
    local $, = "\t"; 
    print "@$factors\n\n";
}

sub show_matches {
    my ($number) = shift;
    local ($matches) = @_;   
	
    print <<'HEADER';
-------
matches
-------

HEADER

    local $ul = '-' x length($number);
	
    formeval('match_number'); 
    write;
    
    formeval('match_matches'); 
    for (local $i = 0; $matches->[$i]; $i++) { write }
    print "\n";
}    

sub formeval {
    my ($ident) = @_;
    
    no warnings 'redefine';
    eval $form{$ident};
    die $@ if $@;
}

BEGIN {
    $form{factors} = '
    format =
@<<<<<<<<<<<<<<<<<<<<<<<<<
$number
@<<<<<<<<<<<<<<<<<<<<<<<<<
$ul
.';

    $form{match_number} = '
    format =
@<<<<<<<<<<<<<<<<<<<<<<<<<
$number
@<<<<<<<<<<<<<<<<<<<<<<<<<
$ul
.';

    $form{match_matches} = '
    format =
@<<<<<<<<<<<* @<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
$matches->[$i][0], $matches->[$i][1]
.';
}


lib/Math/Factor/XS.pm
package Math::Factor::XS;

use strict;
use warnings;
use base qw(Exporter);
use boolean qw(true);

use Carp qw(croak);
use List::MoreUtils qw(all);
use Params::Validate ':all';
use Scalar::Util qw(looks_like_number);

our ($VERSION, @EXPORT_OK, %EXPORT_TAGS);
$VERSION = '0.39_01';
@EXPORT_OK = qw(factors matches prime_factors count_prime_factors);
%EXPORT_TAGS = (all => \@EXPORT_OK);

validation_options(
    on_fail => sub
{
    my ($error) = @_;
    chomp $error;
    croak $error;
});

my $positive_nums = sub
{
    all { looks_like_number($_) && ($_ >= 0) }
      ref $_[0] ? @{$_[0]} : ($_[0]);
};

sub matches
{
    validate_pos(@_,
        { type => SCALAR,
          callbacks => {
            'is a positive number' =>
            $positive_nums,
          },
        },
        { type => ARRAYREF,
          callbacks => {
            'factors are positive numbers' => sub
            {
                my $factors = shift;
                !@$factors or $positive_nums->($factors);
            },
          },
        },
        { type => HASHREF,
          optional => true,
        },
    );
    return xs_matches(@_);
}

require XSLoader;
XSLoader::load(__PACKAGE__, $VERSION);

1;
__END__

=head1 NAME

Math::Factor::XS - Factorize numbers and calculate matching multiplications

=head1 SYNOPSIS

 use Math::Factor::XS ':all';
 # or
 use Math::Factor::XS qw(factors prime_factors matches);

 $number = 30107;

 @factors = factors($number);
 @primes = prime_factors($number);
 @matches = matches($number, \@factors);

 print "$factors[1]\n";
 print "$number == $matches[0][0] * $matches[0][1]\n";

=head1 DESCRIPTION

C<Math::Factor::XS> factorizes numbers by applying trial divisions.

=head1 FUNCTIONS

=head2 factors

Find all factors (ie. divisors) of a number.

 @factors = factors($number);

The number is factorized and its factors are returned as a list.  For
example,

 @factors = factors(30);
 #  @factors = (2, 3, 5, 6, 10, 15);

=head2 prime_factors

Find prime factors of a number.

 @factors = prime_factors($number);

The number is factorized and its prime factors are returned as a list.
Multiplying the list together gives C<$number>.  For example,

 @primes = prime_factors(90);
 #  @primes = (2, 3, 3, 5);

=head2 count_prime_factors

Return the count of prime factors of a number.  This is the number of values
returned by C<prime_factors()>.

 my $count = count_prime_factors($number);

=head2 matches

Calculates matching multiplications.

 @matches = matches($number, \@factors, { skip_multiples => [0|1] });

The factors will be multiplied against each other and all combinations
that equal the number itself will be returned as a two-dimensional list.
The matches are accessible through the indexes; for example, the first
two numbers that matched the number may be accessed by C<$matches[0][0]>
and C<$matches[0][1]>, the second pair by C<$matches[1][0]> and
C<$matches[1][1]>, and so on.

The hashref provided at the end is optional. If C<skip_multiples>
is set to a true value, then matching multiplications that contain
multiplicated small factors will be discarded. Example:

 11 * 2737 == 30107 # accepted
 77 * 391  == 30107 # discarded

Direct use of C<$Math::Factor::XS::Skip_multiple> does no longer
have an effect as it has been superseded by C<skip_multiples>.

=head1 EXPORT

=head2 Functions

C<factors()>, C<matches()> and C<prime_factors()> are exportable.

=head2 Tags

C<:all - *()>

=head1 AUTHOR

Steven Schubiger <schubiger@cpan.org>

=head1 LICENSE

This program is free software; you may redistribute it and/or
modify it under the same terms as Perl itself.

See L<http://dev.perl.org/licenses/>

=cut
